zabbix_smartmontools/__init__.py

```diff
@@ -1,150 +1,204 @@
+            identifier = serial
+        else:
+            identifier = device
+        jsonData.append({'{#DDRIVESTATUS}': identifier})
+        if not serial in reported_serials or not config['skipDuplicates']:
+            jsonData.append({"{#DISKID}": identifier})
+            jsonData.append({"{#DISKIDSAS}": identifier})
+        if serial:
+            reported_serials.add(serial)
+    return jsonData
+
+
+# Get the serial number of a hard drive, using the fastest method available on
+# each platform.
+def getSerial(device):
+    """ Get the serial number of a hard drive """
+    serial = None
+    if 'freebsd' in sys.platform:
+        DISK_IDENT_SIZE = 256       # from sys/disk.h
+        DIOCGIDENT = 0x41006489     # from sys/disk.h
+        ident = DISK_IDENT_SIZE * r" "
+        f = open("/dev/%s" % device, 'r')
+        try:
+            serial = ioctl(f, DIOCGIDENT, ident).decode().strip("\x00")
+        except OSError as e:
+            if e.errno == errno.EBADF:
+                # EBADF indicates something that's not a disk, such as a SES
+                # enclosure
+                True
             else:
-                startStopMaxResult = 'UNKNOWN'
-
-            sender.append('%s smartctl.value[%s,startStopMax] "%s"' % (host, dR, startStopMaxResult))
-
-    defectsRe = re.search(r'^Elements in grown defect list:\s+(\d+)', p, re.M | re.I)
-    if defectsRe:
-        sender.append('%s smartctl.value[%s,defects] "%s"' % (host, dR, defectsRe.group(1)))
-
-    poweredHoursRe = re.search(r'^number of hours powered up \=\s+(\d+)', p, re.M | re.I)
-    if poweredHoursRe:
-        sender.append('%s smartctl.value[%s,poweredHours] "%s"' % (host, dR, poweredHoursRe.group(1)))
-
-    nonMediumErrorsRe = re.search(r'^Non-medium error count:\s+(\d+)', p, re.M | re.I)
-    if nonMediumErrorsRe:
-        sender.append('%s smartctl.value[%s,nonMediumErrors] "%s"' % (host, dR, nonMediumErrorsRe.group(1)))
-
-    if sender:
-        error = False
+                raise e
+    elif sys.platform.startswith('linux'):
+        # TODO: is there a way to get this without using a subprocess?
+        cp = subprocess.run(["/sbin/udevadm", "info", "--query=all",
+            "--name=/dev/%s" % device], capture_output=True)
+        for line in cp.stdout.decode().splitlines():
+            match = re.match("ID_SERIAL_SHORT=(\S+)", line)
+            if match:
+                serial = match.groups()[1]
+                break
     else:
-        error = True
-
-    json.append({'{#DISKIDSAS}':dR})
-
-    return error, sender, json
-
-
-def getSmart(config, host, command, d):
-    #print("d:\t'%s'" % d)
-    d = d.strip()
-    if not 'Disks' in config:
-        d = d.replace('-d scsi', '-d auto')   # prevent empty results
-
-    #print("dS:\t'%s'" % d)
-    dR = clearDiskTypeStr(d)   # sanitize the item key
-    dR = sanitizeStr(dR)
-    dOrig = dR   # save original sanitized device name
+        raise NotImplementedError
```

Comment:
[Windows] Since the device here is sd*, it's necessary to reimplement `smartctl`'s logic of assigning the letters exactly in order to achieve repeatable behavior. It's more suitable to utilize regular routine (`smartctl -i`), at least as first step approach.

Reply:
On Windows, is there anything like "/dev/" or is the device name just "sd*"?

Reply:
It's /dev/sd* and /dev/csmi*. On all my machines csmi is a duplicate. Perhaps it can be dropped (unless specified in the config).
```
>smartctl --scan
/dev/sda -d ata # /dev/sda, ATA device
/dev/sdb -d ata # /dev/sdb, ATA device
/dev/csmi0,0 -d ata # /dev/csmi0,0, ATA device
/dev/csmi0,1 -d ata # /dev/csmi0,1, ATA device
```